lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90CompoundFormat.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.codecs.lucene90;
+
+import java.io.IOException;
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.codecs.CompoundDirectory;
+import org.apache.lucene.codecs.CompoundFormat;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.DataOutput;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexOutput;
+
+/**
+ * Lucene 9.0 compound file format
+ *
+ * <p>Files:
+ *
+ * <ul>
+ *   <li><code>.cfs</code>: An optional "virtual" file consisting of all the other index files for
+ *       systems that frequently run out of file handles.
+ *   <li><code>.cfe</code>: The "virtual" compound file's entry table holding all entries in the
+ *       corresponding .cfs file.
+ * </ul>
+ *
+ * <p>Description:
+ *
+ * <ul>
+ *   <li>Compound (.cfs) --&gt; Header, FileData <sup>FileCount</sup>, Footer
+ *   <li>Compound Entry Table (.cfe) --&gt; Header, FileCount, &lt;FileName, DataOffset,
+ *       DataLength&gt; <sup>FileCount</sup>
+ *   <li>Header --&gt; {@link CodecUtil#writeIndexHeader IndexHeader}
+ *   <li>FileCount --&gt; {@link DataOutput#writeVInt VInt}
+ *   <li>DataOffset,DataLength,Checksum --&gt; {@link DataOutput#writeLong UInt64}
+ *   <li>FileName --&gt; {@link DataOutput#writeString String}
+ *   <li>FileData --&gt; raw file data
+ *   <li>Footer --&gt; {@link CodecUtil#writeFooter CodecFooter}
+ * </ul>
+ *
+ * <p>Notes:
+ *
+ * <ul>
+ *   <li>FileCount indicates how many files are contained in this compound file. The entry table
+ *       that follows has that many entries.
+ *   <li>Each directory entry contains a long pointer to the start of this file's data section, the
+ *       files length, and a String with that file's name.
+ * </ul>
+ */
+public final class Lucene90CompoundFormat extends CompoundFormat {
+
+  /** Extension of compound file */
+  static final String DATA_EXTENSION = "cfs";
+  /** Extension of compound file entries */
+  static final String ENTRIES_EXTENSION = "cfe";
+
```

Comment:
Shall we keep the extensions or use new ones?

Reply:
What would be the motivation? Do you have an idea for improved names?

Reply:
I guess the only benefit is that you can distinguish the files from the new codec vs previous codec by name.

Reply:
We would like to avoid sharing file extensions across file formats, e.g. given that `.liv` is used by live docs today, it could be confusing to use it for stored fields in the future.

However I don't feel the need to change extensions as we iterate on a given file format, even if we changed completely the way it works. This is something we did in Lucene 5.x for postings, moving from `.frq`/`.prox` to `.doc`/`.pos` but this is not a requirement at all, by the way we changed stored fields in Lucene 4.x and kept the same `.fdt` extension even though the new file format had very little in common with the previous one.

I think we can keep the same file extensions here. It's easy enough to read the first bytes of a `cfs`/`cfe` file in case of doubt to check out the name in the codec header.

Reply:
Sounds good to me